src/mesh/forest/tree.cpp

```diff
@@ -0,0 +1,289 @@
+      loc.tree() == my_id,
+      "Trying to add a meshblock to a tree with a LogicalLocation on a different tree.");
+  if (internal_nodes.count(loc)) return -1;
+  if (leaves.count(loc)) return 0;
+
+  std::stack<LogicalLocation> refinement_locs;
+  auto parent = loc.GetParent();
+  for (int l = loc.level() - 1; l >= 0; --l) {
+    refinement_locs.push(parent);
+    if (leaves.count(parent)) break;
+    parent = parent.GetParent();
+  }
+
+  int added = 0;
+  while (!refinement_locs.empty()) {
+    added += Refine(refinement_locs.top(), enforce_proper_nesting);
+    refinement_locs.pop();
+  }
+
+  return added;
+}
+
+int Tree::Refine(const LogicalLocation &ref_loc, bool enforce_proper_nesting) {
+  PARTHENON_REQUIRE(
+      ref_loc.tree() == my_id,
+      "Trying to refine a tree with a LogicalLocation on a different tree.");
+  // Check that this is a valid refinement location
+  if (!leaves.count(ref_loc)) return 0; // Can't refine a block that doesn't exist
+
+  // Perform the refinement for this block
+  std::vector<LogicalLocation> daughters = ref_loc.GetDaughters(ndim);
+  auto gid_parent = leaves[ref_loc].first;
+  leaves.erase(ref_loc);
+  internal_nodes.insert(ref_loc);
+  for (auto &d : daughters) {
+    leaves.insert(std::make_pair(d, std::make_pair(gid_parent, -1)));
+  }
+  int nadded = daughters.size() - 1;
+
+  if (enforce_proper_nesting) {
+    LogicalLocation parent = ref_loc.GetParent();
+    int ox1 = ref_loc.lx1() - (parent.lx1() << 1);
+    int ox2 = ref_loc.lx2() - (parent.lx2() << 1);
+    int ox3 = ref_loc.lx3() - (parent.lx3() << 1);
+
+    for (int k = 0; k < (ndim > 2 ? 2 : 1); ++k) {
+      for (int j = 0; j < (ndim > 1 ? 2 : 1); ++j) {
+        for (int i = 0; i < (ndim > 0 ? 2 : 1); ++i) {
+          LogicalLocation neigh = parent.GetSameLevelNeighbor(
+              i + ox1 - 1, j + ox2 - (ndim > 1), k + ox3 - (ndim > 2));
+          // Need to communicate this refinement action to possible neighboring tree(s)
+          // and trigger refinement there
+          int n_idx =
+              neigh.NeighborTreeIndex(); // Note that this can point you back to this tree
+          for (auto &[neighbor_tree, orientation] : neighbors[n_idx]) {
+            nadded += neighbor_tree->Refine(
+                orientation.Transform(neigh, neighbor_tree->GetId()));
+          }
+        }
+      }
+    }
+  }
+  return nadded;
+}
+
+std::vector<NeighborLocation> Tree::FindNeighbors(const LogicalLocation &loc) const {
+  const Indexer3D offsets({ndim > 0 ? -1 : 0, ndim > 0 ? 1 : 0},
+                          {ndim > 1 ? -1 : 0, ndim > 1 ? 1 : 0},
+                          {ndim > 2 ? -1 : 0, ndim > 2 ? 1 : 0});
+  std::vector<NeighborLocation> neighbor_locs;
+  for (int o = 0; o < offsets.size(); ++o) {
+    auto [ox1, ox2, ox3] = offsets(o);
+    if (std::abs(ox1) + std::abs(ox2) + std::abs(ox3) == 0) continue;
+    FindNeighborsImpl(loc, ox1, ox2, ox3, &neighbor_locs);
+  }
+
+  const int clev = loc.level() - 1;
```

Comment:
Where is this used (given the following return statement)?

Reply:
that is dead code, removed.